ssl/ssl_key_share.cc

```diff
@@ -176,20 +178,550 @@ class X25519KeyShare : public SSLKeyShare {
+    *out_alert = 0;
+    return true;
+  }
+
+  // Because this is a KEM key share, |peer_key| is actually the ciphertext
+  // resulting from the peer encapsulating the shared secret under our public key.
+  // In Finish(), we use our previously generated private key to decrypt
+  // that ciphertext and obtain the shared secret.
+  bool Finish(Array<uint8_t> *out_secret, uint8_t *out_alert,
+              Span<const uint8_t> peer_key) override {
+    // Basic nullptr checks
+    if (!out_alert) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+      return false;
+    }
+
+    // Set alert to internal error by default
+    *out_alert = SSL_AD_INTERNAL_ERROR;
+
+    if (!out_secret) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+      return false;
+    }
+
+    // ctx_ should have been initialized by a previous call to Offer()
+    if (!ctx_) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+      return false;
+    }
+
+    // Retrieve the lengths of the ciphertext and shared secret
+    size_t ciphertext_len = 0;
+    size_t secret_len = 0;
+    if (!EVP_PKEY_encapsulate(ctx_.get(), nullptr, &ciphertext_len, nullptr,
+                              &secret_len)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // Ensure that peer_key is valid
+    if (!peer_key.data() || peer_key.size() != ciphertext_len) {
+      *out_alert = SSL_AD_DECODE_ERROR;
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // |peer_key| is the ciphertext, encapsulating the shared secret,
+    // that the peer generated using our public key. We use our private
+    // key to decapsulate it, and retain the shared secret by writing
+    // it to |out_secret|.
+    uint8_t *ciphertext = (uint8_t *)peer_key.begin();
+    if (!out_secret->Init(secret_len)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_MALLOC_FAILURE);
+      return false;
+    }
+
+    size_t secret_bytes_written = secret_len;
+    if (!EVP_PKEY_decapsulate(ctx_.get(), out_secret->data(),
+                              &secret_bytes_written, ciphertext,
+                              ciphertext_len) ||
+                              secret_bytes_written != secret_len) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // Success; clear the alert
+    *out_alert = 0;
+    return true;
+  }
+
+ private:
+  int nid_;
+  uint16_t group_id_;
+  UniquePtr<EVP_PKEY_CTX> ctx_;
+};
+
+// A HybridKeyShare consists of key shares from two or more component groups,
+// all of which are used to generate a hybrid shared secret.
+// See https://datatracker.ietf.org/doc/html/draft-ietf-tls-hybrid-design.
+  class HybridKeyShare : public SSLKeyShare {
+  public:
+    HybridKeyShare(uint16_t group_id) : group_id_(group_id), hybrid_group_(nullptr) {
+      for (UniquePtr<SSLKeyShare> &key_share : key_shares_) {
+        key_share = nullptr;
+      }
```

Comment:
Any reason we have to clear this state but not in other key share constructors?

Reply:
This was just initializing these values to null in the constructor to avoid them containing uninitialized data. This initialization has been moved to the constructor now in the latest update